kerastuner/tuners/ultraband.py

```diff
@@ -12,28 +12,283 @@
+            values_hash = self._compute_values_hash(values)
+            if values_hash in self._tried_so_far:
+                collisions += 1
+                if collisions > self._max_collisions:
+                    return None
+                continue
+            self._tried_so_far.add(values_hash)
+            break
+        values['tuner/epochs'] = self._epoch_sequence[self._bracket_index]
+        return values
+
+    def _get_num_brackets(self):
+        """Compute the number of brackets based on the scaling factor"""
+        n = 1
+        v = self.min_epochs
+        while v < self.max_epochs:
+            v *= self.factor
+            n += 1
+        return n
+
+    def _get_model_sequence(self):
+        sizes = []
+        size = self.min_epochs
+        for _ in range(self._num_brackets - 1):
+            sizes.append(int(size))
+            size *= self.factor
+        sizes.append(self.max_epochs)
+        sizes.reverse()
+        return sizes
+
+    def _get_epoch_sequence(self):
+        """Compute the sequence of epochs per bracket."""
+        sizes = []
+        size = self.min_epochs
+        for _ in range(self._num_brackets - 1):
+            sizes.append(int(size))
+            size *= self.factor
+        sizes.append(self.max_epochs)
+
+        previous_size = 0
+        output_sizes = []
+        for size in sizes:
+            output_sizes.append(size - previous_size)
+            previous_size = size
+        return output_sizes
 
 
 class UltraBand(tuner_module.Tuner):
+    """Variation of HyperBand algorithm.
+
+    Args:
+        oracle: Instance of Oracle class.
+        hypermodel: Instnace of HyperModel class
+            (or callable that takes hyperparameters
+            and returns a Model isntance).
+        objective: String. Name of model metric to minimize
+            or maximize, e.g. "val_accuracy".
+        max_trials: Int. Total number of trials
+            (model configurations) to test at most.
+            Note that the oracle may interrupt the search
+            before `max_trial` models have been tested.
+        seed: Int. The random seed. If None, it would use a random number.
+        factor: Int. The factor of the change of number of epochs
+            and the number of models per bracket.
+        min_epochs: Int. The minimum number of epochs to train a model.
+        max_epochs: Int. The maximum number of epochs to train a model.
+    """
 
     def __init__(self,
                  hypermodel,
                  objective,
                  max_trials,
+                 seed=None,
+                 factor=3,
+                 min_epochs=3,
+                 max_epochs=10,
                  **kwargs):
-        oracle = UltraBandOracle()
+        oracle = UltraBandOracle(max_trials,
+                                 seed,
+                                 factor,
+                                 min_epochs,
+                                 max_epochs)
```

Comment:
When calling a method with lots of arguments, pass them by name, this will avoid errors.